src/main/java/com/uid2/admin/vertx/service/ServiceService.java

```diff
@@ -203,6 +205,13 @@ private void handleUpdate(RoutingContext rc) {
                     ResponseUtil.error(rc, 400, "invalid parameter: roles");
                     return;
                 }
+                // check that if role is removed, it is no longer in use by service links
+                Set<Role> rolesToRemove = service.getRoles().stream().filter(r -> !roles.contains(r)).collect(Collectors.toSet());
+                List<ServiceLink> serviceLinks = this.serviceLinkProvider.getAllServiceLinks().stream().filter(sl -> sl.getServiceId() == serviceId).collect(Collectors.toList());
+                if (serviceLinks.stream().anyMatch(sl -> !Collections.disjoint(sl.getRoles(), rolesToRemove))) {
```

Comment:
if (serviceLinks.stream().anyMatch(sl -> sl.getRoles().stream().anyMatch(rolesToRemove::contains)))

is simpler and more readable, avoid "not-not" logic as it's confusing